File: vulntest/tier1-single/use-after-free/go/source/main.go
// Tier 1 — use-after-free (Go / unsafe.Pointer variant).
//
// Pure Go GC prevents UAF. The unsafe.Pointer escape valve lets
// developers obtain raw pointers; if the underlying slice is
// resized (which may relocate the backing array), the raw pointer
// becomes dangling.
//
// Knowledge: [[Memory/Knowledge/bhg_unsafe_pointer_patterns]]
// CWE-416 (analog).
package main

import (
	"fmt"
	"unsafe"
)

func main() {
	s := []byte("hello, world")
	// take raw pointer to backing array
	p := unsafe.Pointer(&s[0])

	// grow slice — may reallocate the backing array
	for i := 0; i < 1024; i++ {
		s = append(s, byte(i))
	}

	// sink: read through stale pointer if backing array moved
	stale := *(*byte)(p)
	fmt.Printf("read=%c\n", stale)
}

File: vulntest/tier1-single/use-after-free/rust/source/src/main.rs
// Tier 1 — use-after-free (Rust / unsafe variant).
//
// Rust safe code cannot UAF — borrow checker enforces lifetime.
// Unsafe Box::from_raw + early Box::drop creates a dangling raw
// pointer; subsequent dereference is UAF.

use std::ptr;

struct Resource { value: i32 }

fn main() {
    let b = Box::new(Resource { value: 0xdead_beefu32 as i32 });
    let raw: *mut Resource = Box::into_raw(b);

    unsafe {
        // free
        let _ = Box::from_raw(raw);
        // sink: read after free
        println!("{}", (*raw).value);
    }
}
